subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/DefaultDependencyCoordinateResolverFactory.java

```diff
@@ -0,0 +1,159 @@
+        VersionMappingStrategyInternal versionMappingStrategy
+    ) {
+        this.projectDependencyResolver = projectDependencyResolver;
+        this.versionMappingStrategy = versionMappingStrategy;
+    }
+
+    public ComponentDependencyResolver createComponentResolver(SoftwareComponentVariant variant) {
+        return getDependencyResolutionStrategy(variant).componentResolver;
+    }
+
+    @Override
+    public VariantDependencyResolver createVariantResolver(SoftwareComponentVariant variant) {
+        return getDependencyResolutionStrategy(variant).variantResolver;
+    }
+
+    public DependencyResolvers getDependencyResolutionStrategy(SoftwareComponentVariant variant) {
+        Configuration configuration = null;
+
+        ImmutableAttributes attributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
+        VariantVersionMappingStrategyInternal versionMapping = versionMappingStrategy.findStrategyForVariant(attributes);
+
+        // Fallback to component coordinate mapping if variant mapping is not enabled
+        ComponentDependencyResolver componentResolver;
+        if (versionMapping.isEnabled()) {
+            if (versionMapping.getUserResolutionConfiguration() != null) {
+                configuration = versionMapping.getUserResolutionConfiguration();
+            } else if (versionMapping.getDefaultResolutionConfiguration() != null) {
+                configuration = versionMapping.getDefaultResolutionConfiguration();
+            }
+        }
+
+        if (configuration != null) {
+            componentResolver = new VersionMappingVariantDependencyResolver(projectDependencyResolver, configuration);
+        } else {
+            componentResolver = new MinimalComponentDependencyResolver(projectDependencyResolver);
+        }
+
+        return new DependencyResolvers(new VariantResolverAdapter(componentResolver), componentResolver);
+    }
+
+    private static class DependencyResolvers {
+
+        private final VariantDependencyResolver variantResolver;
+        private final ComponentDependencyResolver componentResolver;
+
+        public DependencyResolvers(VariantDependencyResolver variantResolver, ComponentDependencyResolver componentResolver) {
+            this.variantResolver = variantResolver;
+            this.componentResolver = componentResolver;
+        }
+    }
+
+    /**
+     * Adapts a {@link ComponentDependencyResolver} to a {@link VariantDependencyResolver}
+     * by returning component-precision coordinates.
+     */
+    private static class VariantResolverAdapter implements VariantDependencyResolver {
+
+        ComponentDependencyResolver delegate;
+
+        public VariantResolverAdapter(ComponentDependencyResolver delegate) {
+            this.delegate = delegate;
+        }
+
+        @Nullable
+        @Override
+        public ResolvedCoordinates resolveVariantCoordinates(ExternalDependency dependency, VariantWarningCollector warnings) {
```

Comment:
💭  `warnings` here really feels like it shouldn't be part of the interface.  It should just be a field on the implementation.  There's nothing intrinsic too finding coordinates that suggests this collector is necessary, it's just an implementation detail of this adapter class.

Reply:
`warnings` cannot be part of the implementation, since the factory is not scoped to a particular variant, while the warnings are scoped to a single variant. 

It would be nice if we could do a multi-return here and return a tuple of the coordinates and the warnings, though that is not possible without defining a "Result" type, containing both the coordiantes and the warnings. IMO this is way too much overhead. 

I agree the warnings API isn't the prettiest but I think this is certainly acceptable for now and we can revisit this when we put more thought into how we should handle publication warnings

Reply:
We could probably supply the warnings when creating the variant-scoped resolvers but again I would like to wait here. This was already in the existing interface and the more changes in this PR the more trouble it is to update the prototype that branches off of this.

The review period for the spec is nearing its end so once we push that though we can work on some of this kind of stuff

Reply:
Agree a "Result" type isn't worth doing here.  Refining this in a future PR is fine with me.